### DISCUSSION TITLE
using Filters in Collection

### DISCUSSION THREAD

**Discussion by Mr-Mohammad** [alirezanet/Gridify#115] [Q&A] [answered]
Hello friends,
When I try to filter on an ICollection, I get an error saying that the field cannot be found.
The structure of my object is as follows:

    public class ApplicantDocument 
    {
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string FatherName { get; set; }
        public Gender? Gender { get; set; }
        public ICollection Companies { get; set; }
    }

The filter I am writing is as follows:
ApplicantDocument.Companies.JuridicalNationalCode = 14010701838

**Comment by alirezanet** [comment]
Hi @Mr-Mohammad, 
Please share your code, how are you using gridify? also, your query doesn't look correct! you need to use your property name

**Reply by Mr-Mohammad** [reply]
Hi @alirezanet 
            var res = _dbContext.WorkFlowTask.Include(x => x.workFlow).ThenInclude(x => x.ApplicantDocument).ThenInclude(c=>c.companies).Gridify(query);

object 
public class ApplicantDocument 
{
    public string FirstName { get; set; }
    public string LastName { get; set; }
    public string FatherName { get; set; }
    public Gender? Gender { get; set; }
    public ICollection<Company> Companies { get; set; }
}

public class Company 
{
    public string JuridicalNationalCode { get; set; }
    public string CompanyName { get; set; }
}

**Reply by alirezanet** [reply, marked as answer]
so you are trying to create an expression on the third level which is not easily possible, In these scenarios, you should start your query in reverse order to access the target properties: e.g.:

```csharp
_dbContext.Companies..Gridify(gquery)
_dbContext.ApplicantDocument.Gridify(gquery)
```

or you can use GridifyMapper to map your third-level objects to a string:

e.g:
``` csharp
// how to map the third level objects depends on your model relationship 
var mapper = new GridifyMapper<WorkFlowTask>(true)
         .AddMap("jnc", v => v.ApplicantDocument.Companies.Select(w=> w.JuridicalNationalCode));
    
   var res = _dbContext.WorkFlowTask.Include(x => x.workFlow).ThenInclude(x => x.ApplicantDocument).ThenInclude(c=>c.companies).Gridify(query,mapper ); 
```
this way you can query like this: 
```
jnc = 14010701838
```

**Reply by alirezanet** [reply]
PS. I just realized `Companies` property is a collection, and updated the example, (you can use the Linq select statement).
More info: https://alirezanet.github.io/Gridify/guide/gridifyMapper.html#sub-collections

**Reply by Mr-Mohammad** [reply]
Excellent 
Thank you @alirezanet
